script.tablo/lib/windows/livetv.py

```diff
@@ -0,0 +1,936 @@
+                airing.gridAiring.displayTimeEnd()
+            )
+        except tablo.ConnectionError:
+            xbmcgui.Dialog().ok(T(32161), T(32157).format(tablo.API.device.displayName))
+            return
+
+        kwargs = {
+            'number': airing.gridAiring.number,
+            'background': airing.background,
+            'callback': self.actionDialogCallback,
+            'obj': airing,
+            'start_indicator1': 'new' in airing.qualifiers and 'indicators/qualifier_new_hd.png' or '',
+            'start_indicator2': 'live' in airing.qualifiers and 'indicators/qualifier_live_hd.png' or ''
+        }
+
+        self.setDialogButtons(airing, kwargs)
+
+        if airing.type == 'movie':
+            try:
+                show = airing.gridAiring.getShow()
+                description = show.plot or show.description
+            except:
+                util.ERROR()
+                description = ''
+        else:
+            description = airing.gridAiring.description
+
+        secs = airing.gridAiring.secondsToStart()
+
+        if secs < 1:
+            start = T(32145).format(util.durationToText(secs * -1))
+        else:
+            start = T(32146).format(util.durationToText(secs))
+
+        actiondialog.openDialog(
+            airing.title,
+            info,
+            description,
+            start,
+            **kwargs
+        )
+        # self.updateIndicators()
+
+    def actionDialogCallback(self, obj, action):
+        airing = obj
+        changes = {}
+
+        if action:
+            if action == 'watch':
+                player.PLAYER.playAiringChannel(airing.gridAiring)
+            elif action == 'record':
+                airing.schedule()
+                self.grid.updateChannelAiringData(path=airing.gridAiring.channel['path'])
+            elif action == 'unschedule':
+                airing.schedule(False)
+                self.grid.updateChannelAiringData(path=airing.gridAiring.channel['path'])
+
+            self.updateGridItem(airing, self.getFocusId())
+
+        if airing.gridAiring.ended():
+            secs = airing.gridAiring.secondsSinceEnd()
+            changes['start'] = T(32144).format(util.durationToText(secs))
+        else:
+            self.setDialogButtons(airing, changes)
+
+            secs = airing.gridAiring.secondsToStart()
+
+            if secs < 1:
+                start = T(32145).format(util.durationToText(secs * -1))
+            else:
+                start = T(32146).format(util.durationToText(secs))
+
+            changes['start'] = start
+
+        return changes
+
+
+class EPGXMLGenerator(object):
+    BASE_ID = 100
+    HALF_HOUR_WIDTH = 320
+    ITEMS_PER_ROW = 16
+
+    BASE_XML_PATH = os.path.join(
+        SKIN_PATH, 'resources', 'skins', 'Main', '720p', 'script-tablo-livetv.xml'
+    )
+
+    OUTPUT_PATH = os.path.join(
+        SKIN_PATH, 'resources', 'skins', 'Main', '720p', 'script-tablo-livetv-generated.xml'
+    )
+
+    CHANNEL_BASE_XML = '''
```

Comment:
Hmm not really a big fan of creating XML files like this.+
Perhaps better move this stuff out of the .py files into separate xml files and parsing them with a proper XML parser?